synapse/api/errors.py

```diff
@@ -600,8 +606,12 @@ def cs_error(msg: str, code: str = Codes.UNKNOWN, **kwargs: Any) -> "JsonDict":
 
 
 class FederationError(RuntimeError):
```

Comment:
I'm slightly surprised that FederationError doesn't inherit from CodeReasonException, given that it has both a code and a reason. :shrug:

Reply:
When we split the errors out, seems like it would be good to make the push error use the `CodeMessageException` as the base exception 👍

Reply:
Oops, s/reason/message in my comment above!

Other cleanups I'd love to see (mentioned for interest):
- rename SynapseError to MatrixAPIError? Or something else which clarifies "this is an internal error for Synapse's consumption only". In a lot of places we raise a SynapseError as a shorthand for "return an error response"
- audit where we raise SynapseErrors from. Seems like they ought to be raised by the rest and handler layers? (The classic example is raising NotFoundError from the db layers which then bubbles all the way up to the rest layer.)